### DISCUSSION TITLE
Using xmlbuilder2 in a webworker or Cloudflare Worker?

### DISCUSSION THREAD

**Discussion by kevin-mitchell** [oozcitak/xmlbuilder2#77] [Q&A]
I am using Cloudflare Workers, which is similar to AWS Lambda or Azure Functions but is not the "full" node runtime. As a result, the build process is a bit strange and webpack targets `webworker` - see [this section of the Worker docs if you're at all currious](https://developers.cloudflare.com/workers/cli-wrangler/webpack#sensible-defaults).

> Cloudflare Workers are built to match the Service Worker API, so we set our target to webworker.

The problem I'm facing is that for a reason I can't understand, the code below (specifically `create(...`) triggers an error:

`ReferenceError: window is not defined`

```
...
import { create } from 'xmlbuilder2';
...
...
const newOrderOMSRequest = {
                api: {                  
                    ApiResponse: {
                        Client: '...',
                        ResponseType: 'NEWORDER',
                        RequestData: {
                            CMS: 'BTS',
                            //...
                            //...
                        }
                    }
                }
            };              
            const newOrderOMSRequestDocument = create(newOrderOMSRequest);
...
```

I don't understand why this is happening exactly, but I can only imagine that some execution path within `xmlbuilder2` perhaps checks for the existence of certain node features (?) and if they are not found it falls back to "browser mode" or something? 

I'm wondering if anybody can point me in the right direction, tell me if this is completely hopeless, etc.

Thanks for your time!

**Comment by chpio** [comment]
I'm generating xml files in a SketchApp plugin (it is running in a Safari sandbox, or something?! But not in a "real web browser") and getting the same error thrown: "Can't find variable: window".

EDIT: removing `"browser": "./lib/xmlbuilder2.min.js",` from "node_modules/xmlbuilder2/package.json" made it work. Need to look into webpack build options, maybe there's something to disable that behavior.

EDIT: got it working with this tweak to the webpack config:

```js
    config.resolve.aliasFields = ["sketch"];
    config.resolve.mainFields = ["sketch", "module", "main"];
```

`"sketch"` is a custom field for sketch plugins, i guess. So usually you would use:

```js
    config.resolve.aliasFields = [];
    config.resolve.mainFields = ["module", "main"];
```

**Comment by oozcitak** [comment]
This should be fixed in version `3.0.1`. See: 95109844e899301ca9576a7c2017b4c1be5b5938
